### DISCUSSION TITLE
Simple publisher and subscriber in same node

### DISCUSSION THREAD

**Discussion by kinrep71** [noelex/rclnet#4] [Q&A] [answered]
Hi all,

can I have one more question? I am trying to manage a full-duplex communication between ROS2 foxy and C# application both running in Windows. Is it generally possible to have a simple publisher and subscriber in one node or more threads should be implemented in C# script? I am trying to implement it like that but it is not working form me. 
![image](https://user-images.githubusercontent.com/49516152/230309905-8d9e19a5-b513-40ee-8f87-1a6c07adcb5c.png)

**Comment by noelex** [comment, marked as answer]
Yes it's possible, but you need to run either the send loop or receive loop in a separate coroutine.

In your code, the `await foreach` statement will keep receiving messages until `sub`  is disposed, thus the publisher code will never have a chance to execute.

To send and receive messages concurrently, you can do `await foreach` in another coroutine, e.g.

```csharp
_ = ReceiveMessagesAsync();

using var pub = ...
while (true)
{
    pub.Publish(...);
    await timer.WaitOneAsync();
}

async Task ReceiveMessagesAsync()
{
    using var sub = ...
    await foreach (var msg in sub.ReadAllAsync())
    {
         ...
    }
}

```

**Reply by kinrep71** [reply]
It works for me! Thank you :)
